pkg/controller/populators/clone-populator.go

```diff
@@ -81,6 +88,90 @@ type ClonePopulatorReconciler struct {
 	multiTokenValidator *cc.MultiTokenValidator
 }
 
+var supportedCloneSources = map[string]client.Object{
+	"VolumeSnapshot":        &snapshotv1.VolumeSnapshot{},
+	"PersistentVolumeClaim": &corev1.PersistentVolumeClaim{},
+}
+
+func isSourceReady(obj client.Object) bool {
+	switch obj := obj.(type) {
+	case *snapshotv1.VolumeSnapshot:
+		return cc.IsSnapshotReady(obj)
+	case *corev1.PersistentVolumeClaim:
+		return cc.IsBound(obj)
+	}
+	return false
+}
+
+func addCloneSourceWatches(mgr manager.Manager, c controller.Controller, log logr.Logger) error {
+	getKey := func(kind, namespace, name string) string {
+		return kind + "/" + namespace + "/" + name
+	}
+
+	if err := mgr.GetClient().List(context.TODO(), &snapshotv1.VolumeSnapshotList{}); err != nil {
+		if meta.IsNoMatchError(err) {
+			// Back out if there's no point to attempt watch
+			return nil
+		}
+		if !cc.IsErrCacheNotStarted(err) {
+			return err
+		}
+	}
+
+	indexingKey := "spec.source"
+
+	// Indexing field for clone sources
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &cdiv1.VolumeCloneSource{}, indexingKey, func(obj client.Object) []string {
+		cloneSource := obj.(*cdiv1.VolumeCloneSource)
+		sourceName := cloneSource.Spec.Source.Name
+		sourceNamespace := cloneSource.Namespace
+		sourceKind := cloneSource.Spec.Source.Kind
+
+		if _, supported := supportedCloneSources[sourceKind]; !supported || sourceName == "" {
+			return nil
+		}
+
+		ns := cc.GetNamespace(sourceNamespace, obj.GetNamespace())
+		return []string{getKey(sourceKind, ns, sourceName)}
+	}); err != nil {
+		return err
+	}
+
+	// Generic mapper function for supported clone sources
+	genericSourceMapper := func(sourceKind string) handler.MapFunc {
+		return func(ctx context.Context, obj client.Object) (reqs []reconcile.Request) {
+			var cloneSources cdiv1.VolumeCloneSourceList
+			matchingFields := client.MatchingFields{indexingKey: getKey(sourceKind, obj.GetNamespace(), obj.GetName())}
+
+			if err := mgr.GetClient().List(ctx, &cloneSources, matchingFields); err != nil {
+				log.Error(err, "Failed to list VolumeCloneSources")
+				return nil
+			}
+
+			for _, cs := range cloneSources.Items {
+				reqs = append(reqs, reconcile.Request{NamespacedName: types.NamespacedName{Namespace: cs.Namespace, Name: cs.Name}})
+			}
+			return reqs
+		}
+	}
+
+	// Register watches for all supported clone source types
+	for kind, obj := range supportedCloneSources {
+		if err := c.Watch(source.Kind(mgr.GetCache(), obj,
+			handler.EnqueueRequestsFromMapFunc(genericSourceMapper(kind)),
+			predicate.Funcs{
+				CreateFunc: func(e event.CreateEvent) bool { return true },
+				DeleteFunc: func(e event.DeleteEvent) bool { return false },
+				UpdateFunc: func(e event.UpdateEvent) bool { return isSourceReady(obj) },
```

Comment:
I think this will requeue for any readyToUse snapshot. Instead, what you're after is a single case where readyToUse switches from false to true, so something like https://github.com/kubevirt/containerized-data-importer/blob/ce1e83c3e509d559cb5e61168860ac28d741a37c/pkg/controller/datasource-controller.go#L307-L309

Reply:
I assume that we still want to requeue any volume snapshot update even after it becomes ready? This would only allow for one single update requeue.

Reply:
The controller requeues the request if there's an error/some other reason. You only have to catch the transition

Reply:
ack